src/utils.h

```diff
@@ -57,18 +57,30 @@ StepSize(const double               max_squared_sum,
 //' @param x A Eigen matrix of sparse or dense type.
 //'
 //' @return The maximum squared norm over columns (samples).
-template <typename T>
-inline
 double
-ColNormsMax(const T& x)
+ColNormsMax(const Eigen::SparseMatrix<double>& x,
+            const Eigen::ArrayXd&              x_center_scaled,
+            const bool                         standardize)
 {
   auto m = x.cols();
 
-  double cn_max = 0.0;
-  for (decltype(m) j = 0; j < m; ++j)
-    cn_max = std::max(cn_max, x.col(j).squaredNorm());
+  double norm_max = 0.0;
+  for (decltype(m) j = 0; j < m; ++j) {
+    double norm =
+      standardize ? (x.col(j) - x_center_scaled.matrix()).squaredNorm()
+                  : x.col(j).squaredNorm();
+
+    norm_max = std::max(norm_max, norm);
+  }
 
-  return cn_max;
+  return norm_max;
+}
+
+double
+ColNormsMax(const Eigen::MatrixXd& x,
+            const Eigen::ArrayXd&  x_center_scaled,
+            const bool             standardize) {
+  return x.colwise().squaredNorm().maxCoeff();
```

Comment:
Just so I'm clear - are we not rescaling `X` in the dense case at all? (I know it's outside the scope of this PR)

Reply:
X is already centered in the dense case so no modification is needed here. x_center_scaled is only used whenever X is sparse and we are standardizing. I'll add a comment regarding this.